### PR TITLE
Added watermark Rust example

### PR DESCRIPTION
**Pull request by qrdl**

### CODE REVIEW

File: fastedge-rust-sdk/examples/watermark/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        offset_y: u32,
+        opacity: f32) -> DynamicImage {
+
+    let opacity = if opacity > 1.0 {
+        1.0
+    } else if opacity < 0.0 {
+        0.0
+    } else {
+        opacity
+    };
```
Comment (ruslanti) [inline fastedge-rust-sdk/examples/watermark/src/lib.rs:171]: let opacity = match opacity {
        o if o > 1.0 => 1.0,
        o if o < 0.0 => 0.0,
        _ => opacity
    };

Comment (ruslanti) [reply fastedge-rust-sdk/examples/watermark/src/lib.rs:171]: it looks better and clearer with match

File: fastedge-rust-sdk/examples/watermark/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        Ok((u, h)) => (u, h),
+    };
+
+    /* Actual request to S3 */
+    let s3_req = Request::builder()
+        .method(Method::GET)
+        .uri(signed_url.as_str())
+        .header("Host", host)
+        .body(Body::empty())
+        .unwrap();
```
Comment (ruslanti) [inline fastedge-rust-sdk/examples/watermark/src/lib.rs:76]: iI would recommend to use .expect("error message") instead for unwrap()

File: fastedge-rust-sdk/examples/watermark/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            Err(_) => return Response::builder()    // opacity is not a number
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(Body::from("Invalid opacity value\n")),
+            Ok(v) => {
+                if v < 0.0 || v > 1.0 {     // opacity is not in 0-1.0 range
+                    return Response::builder()
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(Body::from("Invalid opacity value\n"))
+                }
+                v
```
Comment (ruslanti) [inline fastedge-rust-sdk/examples/watermark/src/lib.rs:129]: It is more clear to move if to match statement in this way:
```
Ok(v)  if v < 0.0 || v > 1.0  => {     // opacity is not in 0-1.0 range
                    return Response::builder()
                        .status(StatusCode::INTERNAL_SERVER_ERROR)
                        .body(Body::from("Invalid opacity value\n"))
                },
Ok(v) =>  v,
```

File: fastedge-rust-sdk/examples/watermark/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        0,  // X offset for watermark placement
+        0,  // Y offset for watermark placement
+        opacity);
+
+    // convert resulting image to PNG
+    let mut c = Cursor::new(Vec::new());
+    let _ = result.write_to(&mut c, ImageOutputFormat::Png);
+    c.seek(SeekFrom::Start(0)).unwrap();
+    let mut out = Vec::new();
+    c.read_to_end(&mut out).unwrap();
```
Comment (ruslanti) [inline fastedge-rust-sdk/examples/watermark/src/lib.rs:146]: Here I see 2 allocation of vector and copy of the image bytes from one to other.
Could you test if this code can be optimised to:
`
let mut out = Vec::new();
let mut c = Cursor::new(out);
let _ = result.write_to(&mut c, ImageOutputFormat::Png);
`

Comment (ruslanti) [reply fastedge-rust-sdk/examples/watermark/src/lib.rs:146]: Even so, DynamicImage has a method into_bytes() to return Vec<u8>. But, you need to convert image to desired format first.

Comment (ruslanti) [reply fastedge-rust-sdk/examples/watermark/src/lib.rs:146]: Why output image is not save in the same format as input?

File: fastedge-rust-sdk/examples/watermark/src/lib.rs (hunk 1 above)
Comment (qrdl) [reply fastedge-rust-sdk/examples/watermark/src/lib.rs:172]: changed

File: fastedge-rust-sdk/examples/watermark/src/lib.rs (hunk 2 above)
Comment (qrdl) [reply fastedge-rust-sdk/examples/watermark/src/lib.rs]: changed

File: fastedge-rust-sdk/examples/watermark/src/lib.rs (hunk 3 above)
Comment (qrdl) [reply fastedge-rust-sdk/examples/watermark/src/lib.rs]: yep, cleaner this way, changed

File: fastedge-rust-sdk/examples/watermark/src/lib.rs (hunk 4 above)
Comment (qrdl) [reply fastedge-rust-sdk/examples/watermark/src/lib.rs]: yes, simplified this bit, also changed to use the same format

**Review by ruslanti — APPROVED** [review]

**Pull request merged**